### PR TITLE
Zero when no associated records (fix sorting)

### PR DESCRIPTION
**Pull request by kbrock**
Depends:

- [x] #31 refactor virtual_totals
- [x] #54 refactor virtual_totals

## Overview

`virtual_aggregate` allows us to define a `sum(column)` up front and assign it to a `virtual_attribute`.

This means we define a virtual attribute as an aggregate of a relation, without defining any ruby or sql.

So `Author.total_ratings` the same as `author.books.sum(:ratings)`

## Problem

Calculating the sum of a column when there are no records is surprisingly not straight forward.

ActiveRecord and ruby treat the sum of no rows as 0. While SQL treats it as `null` (`nil` in ruby terms)

## Samples

```ruby
author = Author.create
author.books.size # => 0
author.books.sum(:rating) # => 0

[].sum # => 0
```

```sql
SELECT sum(ratings) AS sum FROM authors;

sum
---
null
```

## Before

There are a few methods for calculating the sum from an association:

- a one off database query
- a relation that is loaded in memory
- an attribute form a `SELECT` clause.

Unfortunately, our code is not consistent on the return values based upon how the data is retrieved:

```ruby
author = Author.create

author.books.total_rating # => 0
author.books.load.total_ratings # => nil
Author.select(:total_ratings).first.total_rating # => nil
```

## After

```ruby
author = Author.create

author.books.total_rating # => 0
author.books.load.total_ratings # => 0
Author.select(:total_ratings).first.total_rating # => 0
```

## Alternative

While I wanted to return `nil` for this value (see #13), I realize that rails really wants the sum to be 0. A few default values (i.e.: `|| 0`) in the internals of active record's `count` make it difficult to change this value. And I don't think we want to change the internals of rails like this.

This PR makes the `virtual_aggregate` return a consistent 0 when summing no records from the database.

### CODE REVIEW

**Comment by kbrock** [comment]
WIP because this needs more discussions

**Comment by himdel** [comment]
> The use cases for displaying the n/a will now only get triggered if the hardware record does not exist. vs before, it could (inconsistently) get that value if there was a hardware and there were no disks. Does this work for you?

So, this sounds less consistent.

The UI case is really ... if we're summing 0 records, return nil, otherwise return the sum.

If this changes it so that we will get 0 for one kind of 0 records (hardware exists, but no disks), but nil for another kind of 0 records (hardware doesn't exist), that means the logic on UI side will be *more* complicated.


So, either we should go with https://github.com/ManageIQ/activerecord-virtual_attributes/pull/13,
or missing hardware should lead to 0 as well.

**Comment by kbrock** [comment]
@himdel actually, this is the consistent one, it returns a 0 for all cases. #13 returns nil most of the time but returns a 0 in the case that we need the 1 off (N+1 style) query.

```ruby
class Vm
  virtual_delegate :used_disk_storage => :hardware, :allow_nil => true # :default => nil
end
class Hardware
  virtual_aggregate :used_disk_storage # :default => 0
end

Vm.new(:hardware => nil).used_disk_storage == nil
Vm.new(:hardware => Hardware.first).used_disk_storage == 0
```

The anomoly is because of the definition of the delegate in Vm, not the virtual aggregate.

File: lib/active_record/virtual_attributes/virtual_total.rb
Comment (NickLaMuro) [reply lib/active_record/virtual_attributes/virtual_total.rb]: I ended up pushing the branch I was working for the above comment for reference:

https://github.com/ManageIQ/activerecord-virtual_attributes/compare/master...NickLaMuro:fix_kbrock_commit_nightmare

Don't expect you to use this (too much snark/shade in it for one thing...), but it was much easier for me to make the changes this way instead of trying to pseudo code this.

Comment (kbrock) [reply lib/active_record/virtual_attributes/virtual_total.rb]: I tried to get close to your commits.

github is displaying them out of order. it should be:

```
cb353c754a - virtual_aggregate remove unnecessary [] || 0
1657a3be06 - better aggregate support
585d64ad4d - virtual_total nil attribute does not exec query
be6861b9a2 - virtual_aggregate use if else
09c9f31d8b - virtual_aggregate nil attribute does not exec query
216390386f - virtual_total: remove blank case
b1862d919e - For sum, coalesce sql values to 0
```

**Comment by kbrock** [comment]
@NickLaMuro I tried to come close to your commits. Let me know if there is magic in any of those commits and I can split it up one more time

github is displaying them out of order. it should be:

```
cb353c754a - virtual_aggregate remove unnecessary [] || 0
1657a3be06 - better aggregate support
585d64ad4d - virtual_total nil attribute does not exec query
be6861b9a2 - virtual_aggregate use if else
09c9f31d8b - virtual_aggregate nil attribute does not exec query
216390386f - virtual_total: remove blank case
b1862d919e - For sum, coalesce sql values to 0
```

File: lib/active_record/virtual_attributes/virtual_total.rb
Comment (kbrock) [reply lib/active_record/virtual_attributes/virtual_total.rb]: good catch

**Comment by kbrock** [comment]
ok, rebased to master
fixed cop about complexity of method

**Comment by kbrock** [comment]
touched up the refactor commit - it changed parens. (a split up commit should just split up - not change anything)

**Comment by kbrock** [comment]
rebased in #30 - it changed the order of the columns. updated head comment in pr

**Comment by kbrock** [comment]
WIP: waiting on the decision to return 0
waiting for the refactor to merge

**Comment by jrafanie** [comment]
> WIP: waiting on the decision to return 0
> waiting for the refactor to merge

@kbrock where are we on this?  I see a refactor was merged after that comment, are we just waiting on a decision on returning 0?

**Comment by kbrock** [comment]
@jrafanie thanks

I need to touch base with @himdel again to discuss this
He needs to understand issues with the current implementation and how to avoid

File: lib/active_record/virtual_attributes/virtual_total.rb
Comment (kbrock) [inline lib/active_record/virtual_attributes/virtual_total.rb:133]: maybe we should add a default value for this

**Comment by kbrock** [comment]
fixed merge. Currently, virtual totals return a nil or 0 based upon whether an association is loaded into memory. This inconsistency is not good.

We would prefer that `[].sum` were `nil`, but it is 0 for sql/rails and 0 for ruby. so we're stuck with returning 0.

We were keeping the existing behavior around because we wanted a nil. but to be honest, it was sporadic and not reliable. So we need to do a double check on nil anyway.

I'm suggesting going forward on this and returning a non optimal 0, but doing so consistently with both rails and virtual totals.

**Comment by himdel** [comment]
:+1:

I expect this will only make the UI show a zero instead of Nothing in a few more places than it should, both in GTLs and textual summaries.

There's no good way to fix that systematically, so all we can really do is wait for issues and add some kind of count-based logic to those places.

**Comment by NickLaMuro** [comment]
My only 2cents with this is to bump the version of this gem when doing it, as mentioned... a year ago?

**Comment by jrafanie** [comment]
I agree with @NickLaMuro, we should bump the version considering this is a breaking change.  Do we anticipate this adversely affecting calculations in reports?  For example, did we discard nils from averages where 0 will greatly drop averages?  Are there other places we should write tests for in manageiq to ensure this change doesn't make unforeseen issues?  @kbrock

**Comment by kbrock** [comment]
@jrafanie thanks for the questions.

We only use virtual aggregates for sums.

I'll add support for averages to see if this is an issue and if it changes anything

**Comment by kbrock** [comment]
@chessbyte The sum of an empty value is non existent. I feel like that is better represented as `nil` rather than as `0`.

But if we wanted to go this route, we'd have to override both ruby and rails. I took the peace keeper path and said "ok, you guys win"

**Comment by jrafanie** [comment]
@kbrock are we just one checkmark from getting this ready for final review/merge?

**Comment by kbrock** [comment]
@jrafanie sigh. yea, I need to track down some data in miq that actually shows this happening.

**Comment by kbrock** [comment]
I created an ems with no host or vms. I verified that this triggered the code in question.

 ems-before-commit

It looks like we currently display 0 rather than n/a for these already. So there would be no change

@jrafanie is this good for you?
@himdel As far as I can see, I think this gives you a 0 in all cases.

---

Here are all the fields that will be affected by our change. But since we default the value to 0 in ruby, the point is moot.
ASIDE: this also fixes a performance bug where we double query the database if the field returns a nil, only to then default to 0 in ruby.

model| field|value
-----|------|-----
ems (infra) | total_vcpus | hosts.sum(:total_vcpus)
ems (infra) | total_memory |hosts.sum(:ram_size)
ems (cloud)  | total_cloud_vcpus |vms.sum(:cpu_total_cores)
ems (cloud)  | total_cloud_memory |vms.sum(:ram_size)
hardware  | used_disk_storage      |disks.sum(:used_disk_storage)
hardware  | allocated_disk_storage |disks.sum(:size)
AggregationMixin(ems, host)  | aggregate_cpu_speed       |host_hardwares.sum(:aggregate_cpu_speed)
AggregationMixin(ems, host)  | aggregate_cpu_total_cores |host_hardwares.sum(:cpu_total_cores)
AggregationMixin(ems, host)  | aggregate_disk_capacity   |host_hardwares.sum(:disk_capacity)
AggregationMixin(ems, host)  | aggregate_memory          |host_hardwares.sum(:memory_mb)
AggregationMixin(ems, host)  | aggregate_physical_cpus   |host_hardwares.sum(:cpu_sockets)
AggregationMixin(ems, host)  | aggregate_vm_cpus         |vm_hardwares.sum(:cpu_sockets)
AggregationMixin(ems, host)  | aggregate_vm_memory       |vm_hardwares.sum(:memory_mb)

**Comment by jrafanie** [comment]
Sorry @kbrock, I didn't see this mention until today.  Are we ready to merge and tag or do we need anything else in and released before we bump the major version? cc @NickLaMuro

**Comment by kbrock** [comment]
@jrafanie we should be all set on this one.

**Review by jrafanie — APPROVED** [review]

**Pull request merged**
